task01/src/Vector.cpp

```diff
@@ -0,0 +1,318 @@
+    m_array = new int[m_capacity];
+
+    // Copy the elements of the other vector's array to the lvalue vector's array
+    for (std::size_t i = 0; i < m_size; ++i) {
+        m_array[i] = other.m_array[i];
+    }
+
+}
+//--------------------------------------------------------------------------------
+/*copy assignment operator*/
+Vector& Vector::operator=(const Vector& other)
+{
+    // Check for self-assignment
+    if (this != &other)
+    {
+        // Resize the lvalue vector's array if necessary
+        if (m_size != other.m_size)
+        {
+            // Allocate a new array with the same size as the other vector's array
+            int* new_array = new int[other.m_size];
+
+            // Copy the elements of the other vector's array to the new array
+            for (std::size_t i = 0; i < other.m_size; ++i) {
+                new_array[i] = other.m_array[i];
+            }
+
+            // Deallocate the old array and update the lvalue vector's member variables
+            delete[] m_array;
+            m_array = new_array;
+            m_size = other.m_size;
+            m_capacity = other.m_size;
+        }
+        else // No need to resize the array
+        {
+            // Copy the elements of the other vector's array to the lvalue vector's array
+            for (std::size_t i = 0; i < m_size; ++i) {
+                m_array[i] = other.m_array[i];
+            }
+        }
+
+    }
+    return *this;
+}
+
+//--------------------------------------------------------------------------------
+/*move assignment operator, receives a temporary rvalue other and moves it to lvalue*/
+Vector& Vector::operator=(Vector&& other) noexcept {
+    /*std::cout << "move operator" << std::endl;*/
+    if (this != &other) {
+        delete[] m_array;
+        m_size = other.m_size;
+        m_capacity = other.m_capacity;
+        m_array = other.m_array;
+        other.m_size = 0;
+        other.m_capacity = 0;
+        other.m_array = nullptr;
+    }
+    return *this;
+}
+
+//--------------------------------------------------------------------------------
+
+void Vector::push_back(const int& value) {
+    // Check if we need to resize the array
+    if (m_size >= m_capacity) {
+        // Double the capacity to reduce the number of reallocations
+        std::size_t new_capacity = m_capacity * 2;
+        int* new_array = new int[new_capacity];
+
+        // Copy the existing elements to the new array
+        for (std::size_t i = 0; i < m_size; i++) {
+            new_array[i] = m_array[i];
+        }
+
+        // Delete the old array and update the pointer and capacity
+        delete[] m_array;
+        m_array = new_array;
+        m_capacity = new_capacity;
+    }
+
+    // Add the new element to the end of the array and update the size
+    m_array[m_size] = value;
+    m_size++;
+}
+
+//--------------------------------------------------------------------------------
+
+void Vector::resize(std::size_t count) {
```

Comment:
This is redundant. You can just give the args in `resize(..)` a default value. 
In this case: 
```cpp
void Vector::resize(size_t count, int value = 0) { ... }
```